suitcase/tiff/__init__.py

```diff
@@ -1,25 +1,256 @@
+        than creating files on disk. See the documentation for the
+        ``directory`` parameter below.
+
+    Parameters
+    ----------
+    gen : generator
+        xpected to yield ``(name, document)`` pairs
+
+    directory : string, Path or Manager.
+        For basic uses, this should be the path to the output directory given
+        as a string or Path object. Use an empty string ``''`` to place files
+        in the current working directory.
+
+        In advanced applications, this may direct the serialized output to a
+        memory buffer, network socket, or other writable buffer. It should be
+        an instance of ``suitcase.utils.MemoryBufferManager`` and
+        ``suitcase.utils.MultiFileManager`` or any object implementing that
+        inferface. See the suitcase documentation (LINK ONCE WRITTEN) for
+        details.
+
+    file_prefix : str, optional
+        The first part of the filename of the generated output files. This
+        string may include templates as in ``{proposal_id}-{sample_name}-``,
+        which are populated from the RunStart document. The default value is
+        ``{uid}-`` which is guaranteed to be present and unique. A more
+        descriptive value depends on the application and is therefore left to
+        the user.
+
+    **kwargs : kwargs
+        kwargs to be passed to ``tifffile.TiffWriter.save``.
+    """
+    def __init__(self, directory, file_prefix='{uid}-', **kwargs):
+
+        if isinstance(directory, (str, Path)):
+            self.manager = suitcase.utils.MultiFileManager(directory)
+        else:
+            self.manager = directory
+
+        self.artifacts = self.manager._artifacts
+        self._stream_names = {}  # maps stream_names to each descriptor uids
+        self._files = defaultdict(dict)  # maps stream_name to field/file dicts
+        self._filenames = {}  # map stream_name to file names of tiff files
+        self._file_prefix = file_prefix
+        self._templated_file_prefix = ''
+        self._kwargs = kwargs
+        self._start_found = False
+
+    def start(self, doc):
+        '''Extracts `start` document information for formatting file_prefix.
+
+        This method checks that only one `start` document is seen and formats
+        `file_prefix` based on the contents of the `start` document.
+
+        Parameters:
+        -----------
+        doc : dict
+            RunStart document
+        '''
+
+        # raise an error if this is the second `start` document seen.
+        if self._start_found:
+            raise RuntimeError(
+                "The serializer in suitcase.tiff expects documents from one "
+                "run only. Two `start` documents where sent to it")
+        else:
+            self._start_found = True
+
+        # format self._file_prefix
+        self._templated_file_prefix = self._file_prefix.format(**doc)
+
+        # return the start document
```

Comment:
For simplicity's sake, I can't think of a reason that we should return the document. `DocumentRouter.__call__` automatically returns the original `(name, doc)` unless the subclass' start/stop/descriptor/etc. method returns `None`. I think we should just let it do that, because a `Serializer` isn't making interesting _changes_ to the doc that it needs to communicate downstream. It is a sink.

Reply:
Soooo, I lost about a day going down this rabbit hole. It appears that the `event_model.DocumentRouter.event` method actually requires that the `event_page` method returns the `event_page` document. As a result I have left all of them on just in case there are other issues like this that are unforeseen.

Reply:
@awalter-bnl and I discussed this on a separate channel, but for anyone following along, this was already addressed by https://github.com/NSLS-II/event-model/pull/41